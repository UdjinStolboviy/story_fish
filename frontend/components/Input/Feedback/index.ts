export { Feedback, ConditionalFeedback } from "./Feedback";
