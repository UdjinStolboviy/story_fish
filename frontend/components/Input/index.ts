export { Input } from "./Input";
export { Feedback, ConditionalFeedback } from "./Feedback";
