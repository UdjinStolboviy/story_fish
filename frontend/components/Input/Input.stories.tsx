import { ComponentStoryObj, ComponentMeta } from "@storybook/react";
import { expect } from "@storybook/jest";
import { screen, userEvent } from "@storybook/testing-library";

import { Input } from "./Input";
import { Feedback } from "./Feedback";

export default {
  title: "Controls/Input",
  component: Input,
  args: {},
} as ComponentMeta<typeof Input>;

export const PrimaryInput: ComponentStoryObj<typeof Input> = {
  play: async ({ args }) => {
    await userEvent.type(screen.getByRole("textbox"), "String");
    // @todo: https://github.com/storybookjs/storybook/issues/16941
    // await expect(args.onChange).toHaveBeenCalledTimes(6);
  },
  args: {
    placeholder: "Your name",
    label: "Name:",
  },
};

export const WithIcon: ComponentStoryObj<typeof Input> = {
  args: {
    icon: "Search",
    placeholder: "Search",
    height: 4,
  },
};

export const WithInvalidFeedback: ComponentStoryObj<typeof Input> = {
  args: {
    placeholder: "Some text",
    feedback: <Feedback isValid={false}>Required!</Feedback>,
  },
  argTypes: {
    feedback: {
      control: false,
    },
  },
};

export const WithValidFeedback: ComponentStoryObj<typeof Input> = {
  args: {
    placeholder: "Some text",
    label: "Text:",
    feedback: <Feedback isValid={true}>Looks good!</Feedback>,
  },
  argTypes: {
    feedback: {
      control: false,
    },
  },
};
