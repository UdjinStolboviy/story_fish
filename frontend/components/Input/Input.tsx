import {
  FC,
  ChangeEventHandler,
  InputHTMLAttributes,
  ReactChild,
  forwardRef,
  ForwardedRef,
} from "react";
import styled from "@emotion/styled";

import { Icon, AvailableIcons } from "@/components/Icon";
import { boxShadow } from "@/components/styles";

type WrapperProps = {
  /** Input height */
  height?: number;
  /** Input width */
  width?: number;
  /** Label visibilty */
  isLabelVisible?: boolean;
  /** Feedback visibilty */
  isFeedbackVisible?: boolean;
};

const Wrapper = styled.label<WrapperProps>`
  display: grid;
  gap: 0.1rem;
  grid-template-areas:
    "label"
    "input"
    "feedback";
  grid-template-rows: ${({ isLabelVisible, isFeedbackVisible }) => {
    if (isLabelVisible && isFeedbackVisible) {
      return "1fr 3fr 1fr";
    } else if (isLabelVisible) {
      return "1fr 4fr 0fr";
    } else if (isFeedbackVisible) {
      return "0fr 4fr 1fr";
    } else {
      return "0fr 1fr 0fr";
    }
  }};
  width: ${({ width }) => width}rem;
  height: ${({ height }) => height}rem;
  color: ${({ theme }) => theme.font.regular};
  font-size: 1rem;
`;

const InputWrapper = styled.div`
  grid-area: input;
  display: flex;
  position: relative;
  align-items: center;
  width: 100%;
  height: 100%;
`;

const StyledInput = styled.input`
  all: unset;
  width: 100%;
  height: 100%;
  border-radius: 1rem;
  font-size: 1.4rem;
  padding: 0 2.6rem 0 1.4rem;
  color: ${({ theme }) => theme.font.regular};
  ${({ theme }) =>
    boxShadow(theme.components.shadow1, theme.components.shadow2, true)}
  &::placeholder {
    color: ${({ theme }) => theme.font.placeholder};
  }
  &:focus {
    ${({ theme }) =>
      boxShadow(theme.components.shadow1, theme.components.shadow2)}
    ~ svg {
      color: ${({ theme }) => theme.font.regular};
      opacity: 1;
    }
  }
`;

const StyledIcon = styled(Icon)`
  position: absolute;
  right: 0.3rem;
  color: ${({ theme }) => theme.font.placeholder};
  opacity: 0.7;
`;

const Label = styled.span`
  grid-area: label;
  padding-left: 1.4rem;
`;

const Feedback = styled(Label)`
  grid-area: feedback;
`;

export type Props = {
  /** Placeholder  */
  placeholder: string;
  /** onChange handler */
  onChange?: ChangeEventHandler<HTMLInputElement>;
  /** Input label */
  label?: string;
  /** Icon */
  icon?: AvailableIcons;
  /** Feedback for input */
  feedback?: ReactChild;
} & WrapperProps;

export const Input: FC<Props & InputHTMLAttributes<HTMLInputElement>> =
  forwardRef(
    (
      { label, height = 7, width = 20, icon, feedback, className, ...rest },
      ref
    ) => (
      <Wrapper
        isLabelVisible={Boolean(label)}
        isFeedbackVisible={Boolean(feedback)}
        height={height}
        width={width}
        className={className}
      >
        <Label>{label}</Label>
        <InputWrapper>
          <StyledInput ref={ref as ForwardedRef<HTMLInputElement>} {...rest} />
          {icon && <StyledIcon name={icon} />}
        </InputWrapper>
        <Feedback role="alert">{feedback}</Feedback>
      </Wrapper>
    )
  );

Input.displayName = "Input";
