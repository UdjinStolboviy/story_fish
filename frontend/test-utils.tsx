import { FC, ReactElement } from "react";
import { configureStore } from "@reduxjs/toolkit";
import { Provider } from "react-redux";
import { render, RenderOptions } from "@testing-library/react";
import { ThemeProvider } from "@emotion/react";

import { rootReducer, RootState } from "./store";
import { Themes } from "./styles/themes";
import { Layout } from "./components/Layout";

type Options = { preloadedState?: RootState } & RenderOptions;

const customRender = (
  ui: ReactElement,
  { preloadedState, ...options }: Options = {}
) => {
  const store = configureStore({ reducer: rootReducer, preloadedState });

  const Wrapper: FC = ({ children }) => (
    <Provider store={store}>
      <ThemeProvider theme={Themes.light}>{children}</ThemeProvider>
    </Provider>
  );
  return render(ui, { wrapper: Wrapper, ...options });
};

const pageRender = (
  ui: ReactElement,
  { preloadedState, ...options }: Options = {}
) => {
  const store = configureStore({ reducer: rootReducer, preloadedState });

  const Wrapper: FC = ({ children }) => (
    <Provider store={store}>
      <Layout>{children}</Layout>
    </Provider>
  );
  return render(ui, { wrapper: Wrapper, ...options });
};

// re-export everything
export * from "@testing-library/react";

// override render method
export { customRender as render, pageRender };
