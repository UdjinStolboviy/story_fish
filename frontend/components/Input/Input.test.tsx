import { ChangeEventHandler } from "react";
import userEvent from "@testing-library/user-event";

import { render, screen } from "@/test-utils";

import { Input } from "./Input";

describe("Input test cases", () => {
  it("Render check", () => {
    const onChange = jest.fn();
    const { asFragment } = render(
      <Input
        onChange={onChange as unknown as ChangeEventHandler<HTMLInputElement>}
        label="Label"
        placeholder="Plaiceholder"
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });
  it("Render check with icon", () => {
    const onChange = jest.fn();
    const { asFragment } = render(
      <Input
        onChange={onChange as unknown as ChangeEventHandler<HTMLInputElement>}
        icon="Search"
        label="Label"
        placeholder="Plaiceholder"
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });
  it("Check onChange callback", () => {
    const onChange = jest.fn();
    render(
      <Input
        onChange={onChange as unknown as ChangeEventHandler<HTMLInputElement>}
        label="Label"
        placeholder="Plaiceholder"
      />
    );
    const element = screen.getByRole("textbox");
    userEvent.type(element, "String");
    expect(onChange).toHaveBeenCalledTimes(6);
  });
});
