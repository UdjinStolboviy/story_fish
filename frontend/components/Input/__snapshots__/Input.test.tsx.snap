// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input test cases Render check 1`] = `
<DocumentFragment>
  <label
    class="css-7e1mv2"
    height="7"
    width="20"
  >
    <span
      class="css-1grri41"
    >
      Label
    </span>
    <div
      class="css-52ovgc"
    >
      <input
        class="css-1bj11r8"
        placeholder="Plaiceholder"
      />
    </div>
    <span
      class="css-1reg724"
      role="alert"
    />
  </label>
</DocumentFragment>
`;

exports[`Input test cases Render check with icon 1`] = `
<DocumentFragment>
  <label
    class="css-7e1mv2"
    height="7"
    width="20"
  >
    <span
      class="css-1grri41"
    >
      Label
    </span>
    <div
      class="css-52ovgc"
    >
      <input
        class="css-1bj11r8"
        placeholder="Plaiceholder"
      />
      <svg
        aria-hidden="true"
        aria-label="Search"
        class="css-mjc9ur"
        data-icon="search"
        fill="currentColor"
        focusable="false"
        height="2rem"
        role="img"
        viewBox="64 64 896 896"
        width="2rem"
      >
        <path
          d="M909.6 854.5L649.9 594.8C690.2 542.7 712 479 712 412c0-80.2-31.3-155.4-87.9-212.1-56.6-56.7-132-87.9-212.1-87.9s-155.5 31.3-212.1 87.9C143.2 256.5 112 331.8 112 412c0 80.1 31.3 155.5 87.9 212.1C256.5 680.8 331.8 712 412 712c67 0 130.6-21.8 182.7-62l259.7 259.6a8.2 8.2 0 0011.6 0l43.6-43.5a8.2 8.2 0 000-11.6zM570.4 570.4C528 612.7 471.8 636 412 636s-116-23.3-158.4-65.6C211.3 528 188 471.8 188 412s23.3-116.1 65.6-158.4C296 211.3 352.2 188 412 188s116.1 23.2 158.4 65.6S636 352.2 636 412s-23.3 116.1-65.6 158.4z"
        />
      </svg>
    </div>
    <span
      class="css-1reg724"
      role="alert"
    />
  </label>
</DocumentFragment>
`;
